### PR TITLE
[Mime] SMimePart - Ignore closed generator when casting body to string or iterable.

### PR DESCRIPTION
**Pull request by michaljusiega**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/symfony/issues/34674
| License       | MIT
| Doc PR        | ~

This is my proposal PR for a closed generator in `symfony/mime`.
Actually, there is no a "rewind" for closed generator - so we must check via `Generator::valid()` method if iterable body as `Generator` is valid and not closed.

This PR fix bug explained in https://github.com/symfony/symfony/issues/34674.

### CODE REVIEW

File: src/Symfony/Component/Mime/Part/SMimePart.php [hunk 1]
```diff
@@ -73,6 +77,12 @@ public function bodyToIterable(): iterable
             return;
         }
 
+        if ($this->body instanceof \Generator && !$this->body->valid()) {
+            yield from [];
```
Comment (jderusse) [inline src/Symfony/Component/Mime/Part/SMimePart.php]: What is the purpose of this empty yield?

File: src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    public function testConstructorWithInvalidIterable()
+    {
+        $iterable = $this->getIterable();
+        // We using this method for close a generator which should throws: Exception : Cannot traverse an already closed generator
+        iterator_to_array($iterable);
+
+        $p = new SMimePart($iterable, 'multipart', 'signed', []);
+        $this->assertEquals('', $p->bodyToString());
+        $this->assertEquals('', implode('', iterator_to_array($p->bodyToIterable())));
```
Comment (jderusse) [inline src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php:53]: Should this throw an exception instead of silently return an empty string?
What if we have the following test

```
 $iterable = $this->getIterable();

$p = new SMimePart($iterable, 'multipart', 'signed', []);
$this->assertEquals('content', implode('', iterator_to_array($p->bodyToIterable())));
$this->assertEquals('', implode('', iterator_to_array($p->bodyToIterable())));
```

This looks weird to me.

File: src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        $this->assertEquals(new Headers(
+            new ParameterizedHeader('Content-Type', 'multipart/signed', ['charset' => 'utf-8'])
+        ), $p->getPreparedHeaders());
+    }
+
+    private function getIterable(): iterable
+    {
+        $f = fopen('php://memory', 'r+', false);
+        fwrite($f, 'content');
+        rewind($f);
```
Comment (jderusse) [inline src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php]: Do we really need to use a file here? Can't we just:

```
private function getIterable(): iterable
{
    yield 'con';
    yield 'tent';
}
```

**Review by jderusse — COMMENTED** [review]
Thanks you for looking into this, 

I think we should not silent the issue, and throw an exception instead.

**Comment by michaljusiega** [comment]
@jderusse Thanks you for review.

First, I chose to silence the issue using empty string because stacktrace indicates an error serializing information in WDT.

Of course, we can undo the changes and throw an exception, but that doesn't make sense because it is already thrown anyway in https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpKernel/Profiler/FileProfilerStorage.php#L175 during serialization basing on https://github.com/symfony/symfony/issues/34674 stacktrace.

So we have to consider whether we are abandoning this PR instead catch any exception in Profiler like:

```
        try {
            if (false === file_put_contents($file, serialize($data), 0, $context)) {
                return false;
            }
        } catch (\Throwable $exception) {
            return false;
        }
```

WDYT? 

I will look on your review after my work ;)

File: src/Symfony/Component/Mime/Part/SMimePart.php (hunk 1 above)
Comment (michaljusiega) [reply src/Symfony/Component/Mime/Part/SMimePart.php]: Hmhmh, generate `Generator` from empty array ?

I don't know, I found this solution on PHP.net - fixed on 'yield;'

File: src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php (hunk 3 above)
Comment (michaljusiega) [reply src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php]: Yes, we can :)

File: src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php (hunk 2 above)
Comment (michaljusiega) [reply src/Symfony/Component/Mime/Tests/Part/SMimePartTest.php:53]: The result always will be `content` because `bodyToIterable` always generate newest object of Generator which is not closed and still valid.

**Comment by nicolas-grekas** [comment]
This looks suspicious to me. Why isn't the generator replaced by its string version?
Where is the generator consumed in the first place? Maybe this place should be the patched one, to ensure the body is turned into a string there.

**Comment by michaljusiega** [comment]
> Where is the generator consumed in the first place? Maybe this place should be the patched one, to ensure the body is turned into a string there.

I can't find where the generator is consumed :/

**Comment by raziel057** [comment]
Why this ticket is closed? I try to encrypt emails following the documentation https://symfony.com/doc/current/mailer.html#encrypting-messages but I encounter the same issue under SF 6.1 / PHP 8.1 so it seems the Bug is still there.

``
php.CRITICAL: Uncaught Exception: Cannot traverse an already closed generator {"exception":"[object] (Exception(code: 0): Cannot traverse an already closed generator at ...\\vendor\\symfony\\mime\\Part\\SMimePart.php:58)"} []
[2022-08-24T09:42:35.901404+02:00] request.CRITICAL: Uncaught PHP Exception Exception: "Cannot traverse an already closed generator" at ...\vendor\symfony\mime\Part\SMimePart.php line 58 {"exception":"[object] (Exception(code: 0): Cannot traverse an already closed generator at ...\\vendor\\symfony\\mime\\Part\\SMimePart.php:58)"} []
``

**Comment by fabpot** [comment]
I think we are looking for the root cause here and nobody has found it yet.

**Comment by supeoens** [comment]
We've just spent two to three days battling this issue, with the stack only leading us back to the Symfony `SMimePart::__sleep()` implementation. Our use case was signing and encrypting a specific `Mailable` in a Laravel 9 app, which we were handling in a `MessageSending` event since the `Mailable` view has to be rendered. We kept running into this when setting the body on the event's `Message` instance to the body from the encrypted instance returned from the Symfony `SMimeEncrypter`. We went as far as implementing our own mail transport, and then we finally figured it out. We had an unrelated `MessageSent` event handler that implemented the `ShouldQueue` interface. You can probably guess where I'm going at this point, but essentially since these events were queued, the event message had already been serialized, and thus the underlying body generator had been closed, causing the event handler to be unable to read from it.

In summary: if you're encountering this, check whether you have event handlers or other types of queue implementations that might try to revive a serialized `SMimePart` with a closed generator. I don't know if this can be solved easily, but I hope this is useful to whoever might want to try. @fabpot
